test/unit/uri.c

```diff
@@ -583,16 +590,195 @@ test_invalid_string_uri_set(void)
 	footer();
 }
 
+#define RFC3986_unreserved "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ-._~"
+
+const struct str_escape escape_testcase[] = {
+	[0] = {
+		.str = "-._~",
+		.escaped = "-._~",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[1] = {
+		.str = "0123456789",
+		.escaped = "0123456789",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[2] = {
+		.str = "abcdefghijklm",
+		.escaped = "abcdefghijklm",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[3] = {
+		.str = "nopqrstuvwxyz",
+		.escaped = "nopqrstuvwxyz",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[4] = {
+		.str = "ABCDEFGHIJKLM",
+		.escaped = "ABCDEFGHIJKLM",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[5] = {
+		.str = "NOPQRSTUVWXYZ",
+		.escaped = "NOPQRSTUVWXYZ",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+	[6] = {
+		.str = "!$&'()*+,;=",
+		.escaped = "%21%24%26%27%28%29%2A%2B%2C%3B%3D",
+		.unreserved = RFC3986_unreserved,
+		.plus = false,
+	},
+};
+
+/**
+ * Builds an array with unreserved characters.
+ * uri.unreserved() implemented as a Lua function,
+ * unreserved_tbl() replaces Lua implementation for testing purposes.
+ */
+static void
+unreserved_tbl(const char *str, unsigned char unreserved[256])
+{
+	for (int i = 0; i < 256; i++)
+		unreserved[i] = 0;
+
+	for (; *str; str++) {
+		unsigned char ch = (unsigned char)*str;
+		unreserved[ch] = 1;
+	}
+}
+
+static void
+test_escape(void)
+{
+	header();
+	plan(lengthof(escape_testcase) * 2);
+	unsigned char unreserved[256];
+	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
+		const char *unescaped = escape_testcase[i].str;
+		const char *escaped = escape_testcase[i].escaped;
+		bool plus = escape_testcase[i].plus;
+		char *dst = xcalloc(strlen(unescaped) * 3, sizeof("char"));
+		unreserved_tbl(escape_testcase[i].unreserved, unreserved);
+		size_t dst_size = uri_escape(unescaped, strlen(unescaped),
+					     dst, unreserved, plus);
+		isnt(dst_size, 0, "escaped string ('%s') length != 0", dst);
```

Comment:
We can check the length not only against zero, but also against expected length.

```diff
diff --git a/test/unit/uri.c b/test/unit/uri.c
index bbb4e2b91..52c8fd5de 100644
--- a/test/unit/uri.c
+++ b/test/unit/uri.c
@@ -658,7 +658,7 @@ static void
 test_escape(void)
 {
 	header();
-	plan(lengthof(escape_testcase) * 2);
+	plan(lengthof(escape_testcase) * 3);
 	unsigned char unreserved[256];
 	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
 		const char *unescaped = escape_testcase[i].str;
@@ -668,7 +668,10 @@ test_escape(void)
 		unreserved_tbl(escape_testcase[i].unreserved, unreserved);
 		size_t dst_size = uri_escape(unescaped, strlen(unescaped),
 					     dst, unreserved, plus);
-		isnt(dst_size, 0, "escaped string ('%s') length != 0", dst);
+		is(dst_size, strlen(escaped),
+		   "escaped string ('%s') length != %ld", dst, strlen(escaped));
+		is(strlen(dst), strlen(escaped),
+		   "escaped string ('%s') length != %ld", dst, strlen(escaped));
 		is(strcmp(escaped, dst), 0, "escape: '%s' == '%s'",
 		   escaped, dst);
 		free(dst);
@@ -681,7 +684,7 @@ static void
 test_unescape(void)
 {
 	header();
-	plan(lengthof(escape_testcase) * 2);
+	plan(lengthof(escape_testcase) * 3);
 	for (unsigned i = 0; i < lengthof(escape_testcase); i++) {
 		const char *unescaped = escape_testcase[i].str;
 		const char *escaped = escape_testcase[i].escaped;
@@ -689,8 +692,12 @@ test_unescape(void)
 		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
 		size_t dst_size = uri_unescape(escaped, strlen(escaped),
 					       dst, decode_plus);
-		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
-		     dst);
+		is(dst_size, strlen(unescaped),
+		   "unescaped string ('%s') length != %ld", dst,
+		   strlen(unescaped));
+		is(strlen(dst), strlen(unescaped),
+		   "unescaped string ('%s') length != %ld", dst,
+		   strlen(unescaped));
 		is(strcmp(dst, unescaped), 0, "unescape: '%s' == '%s'",
 		   unescaped, dst);
 		free(dst);
@@ -748,7 +755,7 @@ static void
 test_unescape_special_cases(void)
 {
 	header();
-	plan(lengthof(unescape_testcase) * 2);
+	plan(lengthof(unescape_testcase) * 3);
 	for (unsigned i = 0; i < lengthof(unescape_testcase); i++) {
 		const char *unescaped = unescape_testcase[i].str;
 		const char *escaped = unescape_testcase[i].escaped;
@@ -756,8 +763,12 @@ test_unescape_special_cases(void)
 		char *dst = xcalloc(strlen(unescaped), sizeof("char"));
 		size_t dst_size = uri_unescape(escaped, strlen(escaped),
 					       dst, decode_plus);
-		isnt(strlen(dst), 0, "unescaped string ('%s') length != 0",
-		     dst);
+		is(dst_size, strlen(unescaped),
+		   "unescaped string ('%s') length != %ld", dst,
+		   strlen(unescaped));
+		is(strlen(dst), strlen(unescaped),
+		   "unescaped string ('%s') length != %ld", dst,
+		   strlen(unescaped));
 		is(strcmp(dst, unescaped), 0, "unescape: '%s' == '%s'",
 		   unescaped, dst);
 		free(dst);
```

BTW, `test_unescape()` doesn't verify a return value of `uri_unescape()`, only `strlen(dst)`.